src/Processors/QueryPlan/Optimizations/optimizeJoin.cpp

```diff
@@ -12,20 +12,43 @@
 #include <Interpreters/IJoin.h>
 #include <Interpreters/HashJoin/HashJoin.h>
 
+#include <Processors/QueryPlan/JoinStepLogical.h>
+#include <Processors/QueryPlan/ReadFromPreparedSource.h>
+#include <Interpreters/FullSortingMergeJoin.h>
+
 #include <Interpreters/TableJoin.h>
+#include <Processors/QueryPlan/CreateSetAndFilterOnTheFlyStep.h>
 
 #include <Common/logger_useful.h>
 #include <Core/Joins.h>
 #include <ranges>
+#include <memory>
+
+
+namespace DB::Setting
+{
+    extern const SettingsMaxThreads max_threads;
+    extern const SettingsUInt64 max_block_size;
+    extern const SettingsUInt64 min_joined_block_size_bytes;
+}
 
 namespace DB::QueryPlanOptimizations
 {
 
-static std::optional<UInt64> estimateReadRowsCount(QueryPlan::Node & node, bool has_filter = false)
+static std::optional<UInt64> estimateReadRowsCount(QueryPlan::Node & node, bool has_filter, Stack & steps_stack)
 {
     IQueryPlanStep * step = node.step.get();
+    steps_stack.push_back({.node = &node});
+    SCOPE_EXIT({ steps_stack.pop_back(); });
+
+    constexpr size_t max_chain_length = 64;
+    if (steps_stack.size() >= max_chain_length)
+        return {};
+
     if (const auto * reading = typeid_cast<const ReadFromMergeTree *>(step))
     {
+        optimizePrimaryKeyConditionAndLimit(steps_stack);
```

Comment:
This is eye catching that `estimate...` method call `optimize...` method.

Reply:
I moved `optimizeJoinLogical` and `convertLogicalJoinToPhysical` to `optimizeTreeSecondPass` https://github.com/ClickHouse/ClickHouse/blob/bff7904498d53f06bf709e8ed3156acd6f6e4b43/src/Processors/QueryPlan/Optimizations/optimizeTree.cpp#L146

So, now it's called after `optimizePrimaryKeyConditionAndLimit` and we do not need to call in in estimate, this estimations will already be calculated

Reply:
Also that leads to the case that we do optimization earlier and may be several time (2 times).